src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentialsJCasCCompatibilityTest.java

```diff
@@ -38,39 +42,59 @@ public class GitHubAppCredentialsJCasCCompatibilityTest {
 
         assertThat(domainCredentials.size(), is(1));
         List<Credentials> credentials = domainCredentials.get(0).getCredentials();
-        assertThat(credentials.size(), is(1));
-
-        Credentials credential = credentials.get(0);
-        assertThat(credential, instanceOf(GitHubAppCredentials.class));
-        GitHubAppCredentials gitHubAppCredentials = (GitHubAppCredentials) credential;
-
-        assertThat(gitHubAppCredentials.getAppID(), is("1111"));
-        assertThat(gitHubAppCredentials.getDescription(), is("GitHub app 1111"));
-        assertThat(gitHubAppCredentials.getId(), is("github-app"));
-        assertThat(gitHubAppCredentials.getPrivateKey(), hasPlainText(GITHUB_APP_KEY));
+        assertThat(credentials.size(), is(7));
+
+        assertGitHubAppCredential(credentials.get(0), "github-app", "GitHub app 1111");
+        assertGitHubAppCredential(
+                credentials.get(1), "old-owner", "", new AccessSpecifiedRepositories("test", List.of()));
+        assertGitHubAppCredential(
+                credentials.get(2), "new-specific-empty", "", new AccessSpecifiedRepositories(null, List.of()));
+        assertGitHubAppCredential(
+                credentials.get(3), "new-specific-owner", "", new AccessSpecifiedRepositories("test", List.of()));
+        assertGitHubAppCredential(
+                credentials.get(4),
+                "new-specific-repos",
+                "",
+                new AccessSpecifiedRepositories("test", List.of("repo1", "repo2")));
+        assertGitHubAppCredential(
+                credentials.get(5),
+                "new-infer-owner",
+                "",
+                new AccessInferredOwner(),
+                DefaultPermissionsStrategy.CONTENTS_READ);
+        assertGitHubAppCredential(
+                credentials.get(6),
+                "new-infer-repo",
+                "",
+                new AccessInferredRepository(),
+                DefaultPermissionsStrategy.CONTENTS_WRITE);
     }
 
     @Test
+    @ConfiguredWithCode("github-app-jcasc-minimal.yaml")
     public void should_support_configuration_export() throws Exception {
-        Sequence credentials = getCredentials();
-        CNode githubApp = credentials.get(0).asMapping().get("gitHubApp");
+        CNode credentials = getCredentials();
 
-        String exported = toYamlString(githubApp)
+        String exported = toYamlString(credentials)
                 // replace secret with a constant value
                 .replaceAll("privateKey: .*", "privateKey: \"some-secret-value\"");
 
         String expected = toStringFromYamlFile(this, "github-app-jcasc-minimal-expected-export.yaml");
 
+        // TODO: CasC plugin incorrectly oversimplifies the YAML export for new-specific-empty by
+        // fully removing the repositoryAccessStrategy configuration because its inner
+        // configuration is all empty, but that means it no longer round-trips.
```

Comment:
Essentially, if you have a `Describable` with multiple possible implementations, and a single implementation whose inner values are nullable or support empty, a configuration like this:

```
outerContext:
  whatever: foo
  describable:
    myDescribable1:
      innerConfiguration: []
```

Will be exported as this:

```
outerContext:
  whatever: foo
```

Which is not good if `describable` is heterogeneous and has multiple possible implementations! We'd like for it to be exported as the following, which _is_ what happens if `myDescribable1` has no nested configuration (I think because it ends up as a `SCALAR` in [`ConfigurationAsCode.toYaml`](https://github.com/jenkinsci/configuration-as-code-plugin/blob/26dbcf29d287cb1009f235ddbcc2ca8e1ad4870c/plugin/src/main/java/io/jenkins/plugins/casc/ConfigurationAsCode.java#L582) instead of a `MAPPING`):

```
outerContext:
  whatever: foo
  describable: myDescribable1
```